From 100e755ec77c9e3325a561012f2e7a1afb7c4517 Mon Sep 17 00:00:00 2001
From: Nikolajs Agafonovs <nikolajs.agafonovs@gmail.com>
Date: Wed, 21 Oct 2026 16:20:08 +0300
Subject: [PATCH] x86_64-asm: xsave, xrstor, xsaveopt, xsavec, xsaves, xrstors
 and their 64-bit forms

OPC_48 for the 64-bit forms, as fxsaveq.  Byte-compared with GNU as.
---
 x86_64-asm.h | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/x86_64-asm.h b/x86_64-asm.h
index 4f3d8ba..d442304 100644
--- a/x86_64-asm.h
+++ b/x86_64-asm.h
@@ -388,6 +388,18 @@ ALT(DEF_ASM_OP1(fstsw, 0xdd, 7, OPC_MODRM | OPC_FWAIT, OPT_EA ))
        problem with TCC, we don't use extended registers.  */
     DEF_ASM_OP1(fxsaveq, 0x0fae, 0, OPC_MODRM | OPC_48, OPT_EA )
     DEF_ASM_OP1(fxrstorq, 0x0fae, 1, OPC_MODRM | OPC_48, OPT_EA )
+    DEF_ASM_OP1(xsave, 0x0fae, 4, OPC_MODRM, OPT_EA)
+    DEF_ASM_OP1(xsave64, 0x0fae, 4, OPC_MODRM | OPC_48, OPT_EA)
+    DEF_ASM_OP1(xrstor, 0x0fae, 5, OPC_MODRM, OPT_EA)
+    DEF_ASM_OP1(xrstor64, 0x0fae, 5, OPC_MODRM | OPC_48, OPT_EA)
+    DEF_ASM_OP1(xsaveopt, 0x0fae, 6, OPC_MODRM, OPT_EA)
+    DEF_ASM_OP1(xsaveopt64, 0x0fae, 6, OPC_MODRM | OPC_48, OPT_EA)
+    DEF_ASM_OP1(xrstors, 0x0fc7, 3, OPC_MODRM, OPT_EA)
+    DEF_ASM_OP1(xrstors64, 0x0fc7, 3, OPC_MODRM | OPC_48, OPT_EA)
+    DEF_ASM_OP1(xsavec, 0x0fc7, 4, OPC_MODRM, OPT_EA)
+    DEF_ASM_OP1(xsavec64, 0x0fc7, 4, OPC_MODRM | OPC_48, OPT_EA)
+    DEF_ASM_OP1(xsaves, 0x0fc7, 5, OPC_MODRM, OPT_EA)
+    DEF_ASM_OP1(xsaves64, 0x0fc7, 5, OPC_MODRM | OPC_48, OPT_EA)
 
     /* segments */
     DEF_ASM_OP2(arpl, 0x63, 0, OPC_MODRM, OPT_REG16, OPT_REG16 | OPT_EA)
